nucleus/deployment/common/src/main/java/fish/payara/deployment/util/URIUtils.java

```diff
@@ -80,4 +105,46 @@ public static HttpURLConnection openHttpConnection(URI uri) throws IOException {
         }
         return httpConnection;
     }
+
+    /**
+     * Downloads {@code uri} to a temporary file.
+     *
+     * @param uri the URI to download.
+     * @return The temporary file.
+     * @throws IOException if an I/O exception occurs.
+     * @throws ClassCastException if the {@code uri} is not a file and does not represent HTTP(S) resource.
+     */
+    public static File convertToFile(URI uri) throws IOException {
+        if (hasFileScheme(uri)) {
+            return new File(uri);
+        }
+        return FileUtils.createTempFile(openHttpConnection(uri).getInputStream(), "app", "tmp");
+    }
+
+    /**
+     * Tests whether resource represented by {@code uri} exists.
+     *
+     * @param uri the URI to test.
+     * @return {@code true} if a resource exists; {@code false} otherwise.
+     * @throws IOException if an I/O error occurs.
+     * @throws ClassCastException if the {@code uri} is not a file and does not represent HTTP(S) resource.
+     */
+    public static boolean exists(URI uri) throws IOException {
+        if (hasFileScheme(uri)) {
+            return Files.exists(Paths.get(uri));
+        }
+        HttpURLConnection connection = openHttpConnection(uri);
```

Comment:
All connections and resources should be properly closed. See javadoc.

> Calling the close() methods on the InputStream or OutputStream of an HttpURLConnection after a request may free network resources associated with this instance but has no effect on any shared persistent connection. Calling the disconnect() method may close the underlying socket if a persistent connection is otherwise idle at that time.

Reply:
using try-with-resources

Reply:
I slightly misunderstand. Should I call `getInputStream()` on `HttpURLConnection` before call `getResponseCode()`?

Reply:
I was thinking rather about the HttpURLConnection and when it is disconnected. It is not Closeable. And yes, all streams should be closed when you don't need them, which seems alright here.

I see even original code wasn't disconnecting, interesting ... i wonder if it had some consequences ... 
Perhaps this is useful explanation: https://stackoverflow.com/a/10117254/375449
Also ...
https://stackoverflow.com/a/40123634/375449
https://docs.oracle.com/javase/tutorial/networking/urls/connecting.html (btw I dislike examples which don't disconnect/close in finally).

I tried to find some usages of the disconnect method ...
![image](https://user-images.githubusercontent.com/302904/118505772-03cb5280-b72d-11eb-996b-fe8c1f942849.png)

Generally it seems that the behavior depends on concrete implementation and JVM options (see also KeepAliveCache).

Reply:
A you suggesting to disable persistent connections?

Reply:
For Open JDK and clones, calling `disconnect()` does not close underlying shared socket for HEAD request, with or without closing inputstream. As I suggesting, socket returns to connection pool for further reuse. 
For GET request, `disconnect()` close underlying socket if and only if `close()` on `InputStream` **was not** called.
Should we close shared socket after getting *each* remote archive?

Reply:
I propose to replace return in this method with 
```java
if (connection.getResponseCode() == HttpURLConnection.HTTP_OK) {
    return true;
}
connection.disconnect();
return false;
```
This should be enough because `getInputStream()` method returns an empty no-op stream in both OpenJDK and OpenJ9